internal/telemetry/client.go

```diff
@@ -389,48 +427,115 @@ func (c *Client) newRequest(t RequestType) *Request {
+			OS:        getOSName(),
+			OSVersion: getOSVersion(),
+			// TODO (lievan): arch, kernel stuff?
 		},
 	}
+
+	header := &http.Header{
+		"Content-Type":               {"application/json"},
+		"DD-Telemetry-API-Version":   {"v2"},
+		"DD-Telemetry-Request-Type":  {string(t)},
+		"DD-Client-Library-Language": {"go"},
+		"DD-Client-Library-Version":  {version.Tag},
+		"DD-Agent-Env":               {c.Env},
+		"DD-Agent-Hostname":          {hostname},
+		"Datadog-Container-ID":       {internal.ContainerID()},
+	}
+	if c.URL == getAgentlessURL() && isAPIKeyValid(c.APIKey) {
+		header.Set("DD-API-KEY", c.APIKey)
+	}
+	client := c.Client
+	if client == nil {
+		client = defaultHTTPClient
+	}
+	return &Request{Body: body,
+		Header:          header,
+		HTTPClient:      client,
+		URL:             c.URL,
+		TelemetryClient: c}
 }
 
-// submit posts a telemetry request to the backend
-func (c *Client) submit(r *Request) error {
-	b, err := json.Marshal(r)
-	if err != nil {
-		return err
+func (r *Request) submit() error {
+	if r.TelemetryClient == nil {
+		return fmt.Errorf("all telemetry requests must be associated with a telemetry client")
 	}
+	retry, err := r._submit()
+	if err == nil {
+		// submitting to the telemetry client's intended
+		// URL succeeded - turn logging back on
+		r.TelemetryClient.logging(true)
+	} else if retry {
+		// retry telemetry submissions in instances where the telemetry client has trouble
+		// connecting with the agent
+		r.TelemetryClient.log("telemetry submission failed, retrying with agentless: %s", err)
+		r.URL = getAgentlessURL()
+		r.Header.Set("DD-API-KEY", defaultAPIKey())
+		_, err := r._submit()
+		if err != nil {
+			r.TelemetryClient.log("retrying with agentless telemetry failed: %s", err)
+		}
+		// turn off logging after a failed submission to avoid spamming the user with
+		// telemetry error messages
+		r.TelemetryClient.logging(false)
+	}
+	return err
+}
 
-	req, err := http.NewRequest(http.MethodPost, c.URL, bytes.NewReader(b))
+// agentlessRetry determines if we should retry a failed a request with
+// by submitting to the agentless endpoint
+func agentlessRetry(req *Request, resp *http.Response, err error) bool {
+	if req.URL == getAgentlessURL() {
+		// no need to retry with agentless endpoint if it already failed
+		return false
+	}
 	if err != nil {
-		return err
+		// we didn't get a response which might signal a connectivity problem with
+		// agent - retry with agentless
+		return true
+	}
+	// Do not retry with the following status codes:
+	// 400 - client side error
+	// 429 - too many requests
+	// TODO - add more
+	doNotRetry := []int{400, 429}
+	for status := range doNotRetry {
+		if resp.StatusCode == status {
+			return false
+		}
 	}
-	req.Header = http.Header{
-		"Content-Type":              {"application/json"},
-		"DD-Telemetry-API-Version":  {"v1"},
-		"DD-Telemetry-Request-Type": {string(r.RequestType)},
+	return true
+}
+
+func (r *Request) _submit() (bool, error) {
```

Comment:
It's an anti-pattern to have special characters in Go function names. Can we just call this something more specific? We should also explain in the godoc what returning true/false for this function means.